### ISSUE TITLE
[pkg/ottl] Normalize replace_all_patterns function behavior

### ISSUE THREAD

**Issue by krokwen** [open-telemetry/opentelemetry-collector-contrib#32896]
### Component(s)

pkg/ottl

### Is your feature request related to a problem? Please describe.

I'm implementing log sanitizing processor based on transformprocessor

One of goals - replace value in all '*token*' fields with 'redacted' string.

First obvious solution is to use `replace_all_patterns(attributes, "key", "*.token.*", "redacted")` - but it will **just rename** the matched keys
If we look into function source, we will found that this behavior is changing if we use converting function in there : `replace_all_patterns(attributes, "key", "*.token.*", "redacted", SHA1)` - Now, the keys are left as is, but value is changed, with one note - the value isn't a sha1 hash of original value, but a hash of 'redacted' string that doesn't make sense.

### Describe the solution you'd like

Split it into two functions 'replace_all_keys' and 'replace_all_values', let them accept also 'key' and 'value' mode and apply them to only keys or only values. E.g:

```
- replace_all_keys(attributes, "key", "*.token.*", "redacted") # this will replace the key with replacement if key pattern matches
- replace_all_keys(attributes, "value", "*.token.*", "redacted") # this will replace the key with replacement if value matches
- replace_all_values(attributes, "key", "*.token.*", "redacted") # this will replace value with replacement if key matches
- replace_all_values(attributes, "value", "*.token.*", "redacted") # this will replace value with replacement if value matches
```
This approach will allow to use any converting functions without affecting the behavior

### Describe alternatives you've considered

My final solution is to apply SHA1 function with replacement format and the replace it one more time with 'redacted' string because it doesn't make sense to store a hashsum of replacement string.
```
- replace_all_patterns(attributes, "key", ".*token.*", "redacted", SHA1, "redacted %s")
- replace_all_patterns(attributes, "value", "^redacted.*", "redacted")
```

**Comment by TylerHelmuth** [comment]
You've come upon something OTTL doesn't do well: conditions on specific items in a dynamic list.

Ultimately the way we wish we could handle this situation is with something like:

`replace_all_patterns(attributes, "values", "redacted") where IsMatch(<key from attributes>, ".*token.*")`

which we cannot express yet.  I think https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/29289 would give us the solution.

Your proposed solution would also work, but I'd rather solve the underlying problem which is that we cannot express how to access a map's key(s) in a condition.

**Comment by TylerHelmuth** [comment]
Side note, that is a very clever work around, one that I would consider a bug. I think we should patch that once we have a real solution for the kind of transformation you need. In my opinion the expected outcome of using `function` or `replacePattern` in `replace_all_patterns` with `mode=key` should be to run the new key value through the function.

**Comment by TylerHelmuth** [comment]
@rnishtala-sumo curious on your thoughts.

**Comment by rnishtala-sumo** [comment]
One thing about the issue reported and maybe this isn't obvious from the [docs](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/ottl/ottlfuncs/README.md#replace_pattern), is that at present it makes sense to use the optional function with matched group/s in the replacement string.

For example:
```
replace_pattern(name, "^kube_([0-9A-Za-z]+_)", "$$1.", SHA256, "k8s.%s")
```
The above hashes a substring from the original.

It then makes sense that one would want to apply the function on the replacement string.

> In my opinion the expected outcome of using function or replacePattern in replace_all_patterns with mode=key should be to run the new key value through the function.

I agree, the function isn't applied to the key at the moment.

**Comment by jordankay13** [comment]
An alternative solution to the one posted is to use the `cache` map:
```
          # Copy all attributes to the cache
          - merge_maps(cache, attributes, "insert")
          # Keep all allowed attributes
          - keep_matching_keys(attributes, "${env:ALLOWED_KEYS_REGEX}")
          # Delete all allowed attributes from the cache
          - delete_matching_keys(cache, "${env:ALLOWED_KEYS_REGEX}")
          # Redact all attributes in the cache
          - 'replace_all_matches(cache, "*", "redacted: https://github.com/......../allowed-keys.txt")'
          # Update key with redaction prefix
          - replace_all_patterns(cache, "key", "^(.*)$$", "redaction.redacted.$$1")
          # Merge redacted attributes from cache back into span attributes
          - merge_maps(attributes, cache, "insert")
```

**Comment by xreip** [comment]
Hi i've ran into similar issue where i try to apply ottl funcs to multiple keys based on regexp with replace_all_patterns. 

My main goal was to normalize some keys, based on regexp, with the ToSnakeCase func. (I've you've a solution for this prob, i'd take it.)

The behavior of the replace_all_patterns is pretty unclear and I had to make lot of tests to understand what would be the outcome of my statements.

Here are my tests: 

```
processors:
  batch:
  transform:
    error_mode: ignore
    log_statements:
      - context: log
        statements:
          - replace_all_patterns(log.attributes, "key", "\\s", "_" )

--> Replace the spaces in keys : ok

processors:
  batch:
  transform:
    error_mode: ignore
    log_statements:
      - context: log
        statements:
          - replace_all_patterns(log.attributes, "key", "(\\w+)\\s\\w+", "$$1", SHA256)

--> Replace the value with SHA256 encoding of $$1

processors:
  batch:
  transform:
    error_mode: ignore
    log_statements:
      - context: log
        statements:
          - replace_all_patterns(log.attributes, "key", "(\\w+)\\s\\w+", "$$1", ToLowerCase)

--> Replace the value with $$1 to lower

processors:
  batch:
  transform:
    error_mode: ignore
    log_statements:
      - context: log
        statements:
          - replace_all_patterns(log.attributes, "key", "(\\w+)\\s\\w+", SHA256("$$1"))

--> Replace the key with SHA256 encoding : expected behavior

processors:
  batch:
  transform:
    error_mode: ignore
    log_statements:
      - context: log
        statements:
          - replace_all_patterns(log.attributes, "key", "(\\w+)\\s\\w+", ToLowerCase("$$1"))

--> Replace the key with $$1 but not lower-cased...
```

I cannot understand why  SHA256("$$1") works as expected but  ToLowerCase("$$1") doesn't... 

Thanks

Otelcol-contrib v0.121.0

**Comment by panys9** [comment]
@xreip I faced the same and created a PR (#38626) that will make it work. But as you say it is not clear if it is intended behaviour or not.

**Comment by rnishtala-sumo** [comment]
@panys9 @xreip Yes I agree that there's an issue with the `key` replacement in `replace_all_patterns`. I opened a similar PR earlier
https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/32945/files

I think your PR - #38626  is the way to go
